File: 2023/day5/main.py
from typing import List, Tuple
from itertools import groupby, chain
from functools import reduce

Seeds = List[int]
SeedRange = Tuple[int, int]
RangeMap = Tuple[int, int, int]
Map = List[RangeMap]

def read_input(path: str) -> List[str]:
    with open(path, mode="r") as f:
        out = f.readlines()
    return [x.replace("\n", "") for x in out]


def parse_map(s: List[str]) -> Map:
    values = [x.split() for x in s[1:]]
    # Restructure so we have source start index, source end index + what we need to
    # increment by to get destination
    values = [(int(s), int(s) + int(n) - 1, int(d) - int(s)) for (d, s, n) in values]
    values.sort()
    return values


def parse_input(input: str) -> Tuple[Seeds, List[Map]]:
    seeds = input[0].split(": ")[1].split(" ")
    seeds = [int(s) for s in seeds]

    maps = input[2:]
    maps = [
        parse_map(list(group))
        for k, group in groupby(maps, lambda x: x == "")
        if not k
    ]
    return (seeds, maps)


def apply_map(x: int, map: Map) -> int:
    for start, end, increment in map:
        if x >= start and x <= end:
            return x + increment
    return x


# General stratery here is to recursively split ranges where they overlap
# with ranges in the mapping
def apply_map2(x: SeedRange, map: Map) -> List[SeedRange]:
    s, e = x
    for start, end, increment in map:
        # Has overlap
        if s <= end and e >= start:
            # subspell
            if s >= start and e <= end:
                return [(s + increment, e + increment)]
            elif e <= end:
                return apply_map2((s, start - 1), map) + [
                    (start + increment, e + increment)
                ]
            elif s >= start and e > end:
                return [(s + increment, end + increment)] + apply_map2(
                    (end + 1, e), map
                )
            # s < start and e > end:
            else:
                # as maps are sorted we know there is no overlap for the first part
                return [
                    (s, start - 1),
                    (start + increment, end + increment),
                ] + apply_map2((end + 1, e), map)
    return [x]


def map_all_seed_ranges(seed_ranges: List[SeedRange], map: Map) -> List[SeedRange]:
    out = []
    for x in seed_ranges:
        out += apply_map2(x, map)
    return out


def main() -> None:
    input = read_input("input.txt")
    seeds, maps = parse_input(input)
    locations = [reduce(apply_map, maps, s) for s in seeds]

    part1 = min(locations)
    print(f"Part 1: {part1}")

    n_ranges = len(seeds) // 2
    seed_ranges = [(seeds[2 * i], seeds[2 * i + 1]) for i in range(0, n_ranges)]
    seed_ranges = [(s, s + n - 1) for s, n in seed_ranges]

    locations_ranges = reduce(map_all_seed_ranges, maps, seed_ranges)
    part2 = min(l for l in chain(*locations_ranges))
    print(f"Part 2: {part2}")

if __name__ == "__main__":
    main()

File: 2023/day5/src/main.rs
use std::fs::read_to_string;

fn read_input(filename: &str) -> Vec<String> {
    read_to_string(filename)
        .unwrap()
        .lines()
        .map(String::from)
        .collect()
}

#[derive(Debug, Clone, Copy)]
struct RangeMap {
    start: i64,
    end: i64,
    increment: i64,
}

#[derive(Debug, Clone, Copy)]
struct Range {
    start: i64,
    end: i64,
}

type Map = Vec<RangeMap>;
type Seeds = Vec<i64>;

fn parse_seeds(input: &[String]) -> Seeds {
    input
        .get(0)
        .unwrap()
        .split(": ")
        .nth(1)
        .unwrap()
        .split(' ')
        .map(|x| x.parse::<i64>().unwrap())
        .collect()
}

fn parse_maps(input: &[String]) -> Vec<Map>{
    let mut maps_input = input
        .iter()
        .skip(3);

    // TODO: figure out how group_by works
    let mut map: Vec<RangeMap> = Vec::new();
    let mut maps: Vec<Vec<RangeMap>> = Vec::new();
    while let Some(x) = maps_input.next() {
        if x.is_empty() {
            maps.push(map);
            map = Vec::new();
            // Skip label
            maps_input.next();
            continue;
        }
        let mut numbers = x.split(' ');
        let dest = numbers.next().unwrap().parse::<i64>().unwrap();
        let source = numbers.next().unwrap().parse::<i64>().unwrap();
        let len = numbers.next().unwrap().parse::<i64>().unwrap();
        let range_map = RangeMap{start: source, end: source + len, increment: dest - source,};
        map.push(range_map);
    }
    maps.push(map);
    maps
}

fn apply_map(x: i64, map: &Map) -> i64 {
    
    for range_map in map {
        if x >= range_map.start && x <= range_map.end {
            return x + range_map.increment;
        }
    }
    x
}

fn apply_map2(seed: &Range, map: &Map) -> Vec<Range> {
    let mut queue: Vec<Range> = vec![seed.clone()];
    let mut out: Vec<Range> = Vec::new();
    while let Some(s) = queue.pop() {
        let mut found = false;
        for m in map.iter() {
            // Overlap
            if s.start <= m.end && s.end >= m.start {
                found = true;
                if s.start >= m.start && s.end <= m.end {
                    out.push(Range{start: s.start + m.increment, end: s.end + m.increment});
                } else if s.end <= m.end {
                    out.push(Range{start: m.start + m.increment, end: s.end + m.increment});
                    queue.push(Range{start: s.start, end: m.start - 1});
                } else if s.start >= m.start && s.end > m.end {
                    out.push(Range{start: s.start + m.increment, end: m.end + m.increment});
                    queue.push(Range{start: m.end + 1, end: s.end})
                } else {
                    out.push(Range{start: m.start + m.increment, end: m.end + m.increment});
                    queue.push(Range{start: s.start, end: m.start - 1});
                    queue.push(Range{start: m.end + 1, end: s.end});
                }
            }
        }
        if !found {
            out.push(s);
        }
    }
    out
}

fn map_all_ranges(seed_ranges: Vec<Range>, map: &Map) -> Vec<Range> {
    seed_ranges
        .iter()
        .flat_map(|s| apply_map2(s, map))
        .collect()
}

fn main() {
    
    let input = read_input("input.txt");
    let seeds = parse_seeds(&input);
    let maps = parse_maps(&input);

    let part1 = seeds
        .iter()
        .map(|&x| maps.iter().fold(x, apply_map)) 
        .min()
        .unwrap();
    
    println!("Part 1: {}", part1);

    let starts = seeds.iter().step_by(2);
    let ends = seeds.iter().skip(1).step_by(2);
    let seed_ranges: Vec<Range> = starts
        .zip(ends)
        .map(|(start, end)| Range{start: *start, end: *start + *end - 1})
        .collect();

    let part2 = maps
        .iter()
        .fold(seed_ranges, map_all_ranges)
        .iter()
        .map(|r| r.start)
        .min()
        .unwrap();

    println!("Part 2: {:?}", part2);

}
